contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,286 @@
+
+        require(move < 3, INVALID_MOVE_MSG);
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_TWO_MOVE, INVALID_STEP_MSG);
+        address playerOne = game.playerOne;
+        require(msg.sender == playerOne, ONLY_PLAYER_ONE_CALL_MSG);
+
+
+        //Validate move
+        bytes32 expectedMoveHash = createGameKey(secret, move);
+        require(gameKey == expectedMoveHash, "Move and secret do not match");
+
+        game.playerOneMove = move;
+        game.step = Steps.PLAYER_ONE_REVEAL;
+    }
+
+    /**
+    * Once the opponent has made their move, the creator of the game must resolve to determine the winner
+    * and reward the winner.
+    * If the creator does not resolve the game then the opponent can collect their winnings
+     * after the forfeit window has expired.
+    */
+    function playerTwoReveal(bytes32 gameKey, bytes32 secret, uint8 playerTwoMove) external whenNotPaused {
+
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(game.step == Steps.PLAYER_ONE_REVEAL, INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+        require(msg.sender == playerTwo, INVALID_PLAYER_MSG);
+        require(playerTwoMove < 3, INVALID_MOVE_MSG);
+
+        //Validate move
+        bytes32 expectedMoveHash = createOpponentMoveHash(gameKey, secret, playerTwoMove);
+        require(game.playerTwoMoveHash == expectedMoveHash, "Move and secret do not match");
+
+        uint8 playerOneMove = game.playerOneMove;
+
+
+        //Draw
+        if(playerOneMove == playerTwoMove){
+            resetGame(game);
+            balances[playerOne] = balances[playerOne].add(stake);
+            balances[playerTwo] = balances[playerTwo].add(stake);
+            emit Draw(gameKey);
+            return;
+        }
+
+        //Determine winner
+        address winner;
+        address loser;
+        uint256 winnings = SafeMath.mul(stake, 2);
+
+        // If the creator wins then is rewarded by not paying a fee on the winnings
+        if (isWinner(playerOneMove, opponentMove)) {
+            winner = playerOne;
+            loser = playerTwo;
+        } else {
+            winner = playerTwo;
+            loser = playerTwo;
+        }
+        balances[winner] = balances[winner].add(winnings);
+        resetGame(game);
+
+        emit WinnerRewarded(gameKey, winner, amount);
+    }
+
+    function isWinner(uint8 creatorMove, uint8 opponentMove) internal pure returns (boolean) {
+        return (mod(opponentMove.add(1), 3) == creatorMove);
+    }
+
+    /*
+    * If expiry has passed, the opponent forfeits their stake
+    * They also pay no fee as a compensation
+    * Creator punished by not getting a fee refunded that would have received if resolved
+    */
+    function playerOneCollectsForfeit(bytes32 gameKey) whenNotPaused external whenNotPaused {
```

Comment:
Do you see how player1 can collect forfeit 95% of the time?

Reply:
Even with this?

` require(game.step == uint8(Steps.PLAYER_ONE_REVEAL), INVALID_STEP_MSG);`

`require(block.timestamp >= game.expiryDate, GAME_NOT_EXPIRED_MSG);`

My reasoning is that by the expiry time, player two will have already have sent their hashed move written to the block but won't yet have revealed.

Reply:
Actually I can see the logic here in this function is faulty. That is the consequence of having no tests!

Reply:
My 95% comment is not because the logic in this function is faulty, it is because of how you use the deadline, even before I look at the code.

Reply:
I think I know why. I will write tests. This is very much scaffolding and the tests will hopefully fix identify these holes.